Mathlib/Order/BooleanSubalgebra.lean

```diff
@@ -0,0 +1,381 @@
+@[simp, norm_cast] lemma val_himp (a b : L) : a ⇨ b = (a : α) ⇨ b := rfl
+
+@[simp] lemma mk_bot : (⟨⊥, bot_mem⟩ : L) = ⊥ := rfl
+@[simp] lemma mk_top : (⟨⊤, top_mem⟩ : L) = ⊤ := rfl
+@[simp] lemma mk_sup_mk (a b : α) (ha hb) : (⟨a, ha⟩ ⊔ ⟨b, hb⟩ : L) = ⟨a ⊔ b, L.supClosed ha hb⟩ :=
+  rfl
+@[simp] lemma mk_inf_mk (a b : α) (ha hb) : (⟨a, ha⟩ ⊓ ⟨b, hb⟩ : L) = ⟨a ⊓ b, L.infClosed ha hb⟩ :=
+  rfl
+@[simp] lemma compl_mk (a : α) (ha) : (⟨a, ha⟩ : L)ᶜ = ⟨aᶜ, compl_mem ha⟩ := rfl
+@[simp] lemma mk_sdiff_mk (a b : α) (ha hb) : (⟨a, ha⟩ \ ⟨b, hb⟩ : L) = ⟨a \ b, sdiff_mem ha hb⟩ :=
+  rfl
+@[simp] lemma mk_himp_mk (a b : α) (ha hb) : (⟨a, ha⟩ ⇨ ⟨b, hb⟩ : L) = ⟨a ⇨ b, himp_mem ha hb⟩ :=
+  rfl
+
+/-- A boolean subalgebra of a lattice inherits a boolean algebra structure. -/
+instance instBooleanAlgebraCoe (L : BooleanSubalgebra α) : BooleanAlgebra L :=
+  Subtype.coe_injective.booleanAlgebra _ val_sup val_inf val_top val_bot val_compl val_sdiff
+    val_himp
+
+/-- The natural lattice hom from a boolean subalgebra to the original lattice. -/
+def subtype (L : BooleanSubalgebra α) : BoundedLatticeHom L α where
+  toFun := ((↑) : L → α)
+  map_bot' := L.val_bot
+  map_top' := L.val_top
+  map_sup' := val_sup
+  map_inf' := val_inf
+
+@[simp, norm_cast] lemma coe_subtype (L : BooleanSubalgebra α) : L.subtype = ((↑) : L → α) := rfl
+lemma subtype_apply (L : BooleanSubalgebra α) (a : L) : L.subtype a = a := rfl
+
+lemma subtype_injective (L : BooleanSubalgebra α) : Injective <| subtype L := Subtype.coe_injective
+
+/-- The inclusion homomorphism from a boolean subalgebra `L` to a bigger boolean subalgebra `M`. -/
+def inclusion (h : L ≤ M) : BoundedLatticeHom L M where
+  toFun := Set.inclusion h
+  map_bot' := rfl
+  map_top' := rfl
+  map_sup' _ _ := rfl
+  map_inf' _ _ := rfl
+
+@[simp] lemma coe_inclusion (h : L ≤ M) : inclusion h = Set.inclusion h := rfl
+lemma inclusion_apply (h : L ≤ M) (a : L) : inclusion h a = Set.inclusion h a := rfl
+
+lemma inclusion_injective (h : L ≤ M) : Injective <| inclusion h := Set.inclusion_injective h
+
+@[simp] lemma inclusion_rfl (L : BooleanSubalgebra α) : inclusion le_rfl = .id L := rfl
+@[simp] lemma subtype_comp_inclusion (h : L ≤ M) : M.subtype.comp (inclusion h) = L.subtype := rfl
+
+/-- The maximum boolean subalgebra of a lattice. -/
+instance instTop : Top (BooleanSubalgebra α) where
+  top.carrier := univ
+  top.bot_mem' := mem_univ _
+  top.compl_mem' _ := mem_univ _
+  top.supClosed' := supClosed_univ
+  top.infClosed' := infClosed_univ
+
+/-- The trivial boolean subalgebra of a lattice. -/
+instance instBot : Bot (BooleanSubalgebra α) where
+  bot.carrier := {⊥, ⊤}
+  bot.bot_mem' := by aesop
+  bot.compl_mem' := by aesop
+  bot.supClosed' _ := by aesop
+  bot.infClosed' _ := by aesop
+
+/-- The inf of two boolean subalgebras is their intersection. -/
+instance instInf : Min (BooleanSubalgebra α) where
+  min L M :=  { carrier := L ∩ M
```

Comment:
Do you want to do `min.carrier` etc here too, like above, for consistency? (and because it looks prettier imo)

Reply:
I can't! The type of `min` is `BooleanSubalgebra α → BooleanSubalgebra α → BooleanSubalgebra α`, not `BooleanSubalgebra α`